### PR TITLE
[MLIR] Introduce catalyst.empty, -ones, -zeros supporting dynamic shapes 

### PR DESCRIPTION
**Pull request by grwlf**
In this PR we implement catalyst version of np.empty, ones and zeros, dropping the requirement of knowing shapes at the compile time. All three functions return tensors of dynamic shape.
 
[sc-47631]

### CODE REVIEW

File: frontend/catalyst/jax_primitives.py [hunk 1]
```diff
@@ -1455,6 +1457,43 @@ def _adjoint_lowering(
...
+    output_type = [mlir.aval_to_ir_types(a)[0] for a in jax_ctx.avals_out][0]
+
+    init_value = initializer if not empty else 0
+    initializer_attr = None
+    if finfo_or_none(dtype) is not None:
+        initializer_attr = ir.FloatAttr.get(output_type.element_type, float(init_value))
+    elif iinfo_or_none(dtype) is not None:
+        initializer_attr = ir.IntegerAttr.get(output_type.element_type, int(init_value))
+    else:
+        raise ValueError(f"Unsupported initializer {initializer} of type {dtype}")
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_primitives.py:1491]: Would a ComplexType also make sense here?

**Comment by dime10** [comment]
I know that this implementation might be superseded by builtin JAX support, but I do have a question. Why add all the machinery for a new mlir op and lowering pass, when there are existing operations that can generate tensors?
Was this easier than generating the corresponding MLIR ops from Python?

**Comment by grwlf** [comment]
Closing in favor of https://github.com/PennyLaneAI/catalyst/pull/366

File: frontend/catalyst/jax_primitives.py (hunk 1 above)
Comment (grwlf) [reply frontend/catalyst/jax_primitives.py:1491]: Hmm. I thought we don't have support of complex arrays, do we? I need to check..

**Comment by grwlf** [comment]
> I know that this implementation might be superseded by builtin JAX support, but I do have a question. Why add all the machinery for a new mlir op and lowering pass, when there are existing operations that can generate tensors? Was this easier than generating the corresponding MLIR ops from Python?

@dime10 Thanks for the comment! Indeed, now I think that adding a Catalyst dialect primitive would be an overkill, one would do the same with the Jax primitive alone.

**Pull request closed without merging**